### PR TITLE
chore(permanent_locker): migrate data from legacy to basilisk locker

### PR DESCRIPTION
**Pull request by inventvenkat**
Legacy to basilisk-hs data migration script and an api to test the script
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD
- [x] Data Migration

## Description

Added a script to migrate data from legacy locker to basilisk-hs, which does the following
Step 1: Get all merchants
Step 2: Traverse through all the merchants and get all the payment_methods for each merchant
Step 3: Traverse through all the payment_methods 
Step 4: Call get card api of legacy_locker and call add card api of basilisk-hs 

And an api to test the script
Since we cannot test this script locally, as it needs legacy locker and basilisk to setup locally, 
Step 4 is exposed as an api to verify if that is working.

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [x] This PR modifies application configuration/environment variables

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code
- [x] I added unit tests for my changes where possible
- [x] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

**Review by NishantJoshi00 — CHANGES_REQUESTED** [review]

File: crates/router/src/routes/migrate_legacy_to_basilisk.rs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        .expect("Failed to fetch merchant account from db");
+    let payment_method = state
+        .store
+        .find_payment_method_by_customer_id_merchant_id_list(
+            request_data.customer_id.as_str(),
+            request_data.merchant_id.as_str(),
+        )
+        .await
+        .expect("Failed to fetch payment method from db")
+        .pop()
```
[the remark this answers is not in the document]
Comment (inventvenkat) [reply crates/router/src/routes/migrate_legacy_to_basilisk.rs:35]: routes::migrate_legacy_to_basilisk method is just for testing the migration. I'll rename the function to make it clearer

File: crates/storage_models/src/query/merchant_account.rs [hunk 2]
```diff
@@ -90,4 +90,14 @@ impl MerchantAccount {
...
+
+    pub async fn find_all_merchants(conn: &PgPooledConn) -> StorageResult<Vec<Self>> {
+        generics::generic_filter::<
+            <Self as HasTable>::Table,
+            _,
+            <<Self as HasTable>::Table as Table>::PrimaryKey,
+            _,
+        >(conn, dsl::merchant_id.is_not_null(), None, None, None)
+        .await
+    }
```
[the remark this answers is not in the document]
Comment (inventvenkat) [reply crates/storage_models/src/query/merchant_account.rs:102]: I am removing this method and adding find_all_payment_methods, which has ~1000 records and gonna read it from read_replica

File: crates/router/src/bin/router.rs [hunk 3]
```diff
@@ -36,6 +43,49 @@ async fn main() -> ApplicationResult<()> {
 
     let _guard = logger::setup(&conf.log);
 
+    #[cfg(feature = "migrate_data_from_legacy_to_basilisk_hs")]
+    {
+        let (tx, _rx) = oneshot::channel();
+        let state = routes::AppState::new(conf.clone(), tx).await;
+        let (tx, _rx) = oneshot::channel();
+        let store = Store::new(&conf, true, tx).await;
```
[the remark this answers is not in the document]
Comment (inventvenkat) [reply crates/router/src/bin/router.rs:51]: We cannot get store out of state, because it is inside dynamic dispatch

**Comment by SanchithHegde** [comment]
Closing this as not planned.

**Pull request closed without merging**